쥬스메이커 [STEP2] Lingo, Taeangel

Taeangel opened the pull request:
@jungseungyeo 
안녕하세요 린생 😃
9_group의 Lingo, Taeangel 입니다. 잘 부탁드립니다!
쥬스 메이커 [STEP 2] PR 보냅니다.

## 고민한 점

- 메인 스토리보드에서 재고관리 하는 부분에 Navigation Controller이 기본으로 설정되어 있는 이유가 궁금했고 이 이유가 Navigation bar를 활용하기 위해서 기본으로 설정되어 있다고 생각했습니다. 

- 함수를 연속적으로 사용하도록 구성해서 결합도가 높아지는 것을 해결할 방법에 대해 고민했습니다. 

- 화면에 보이는 Stock과 실제 Stock를 연결하는 방법에 대해 궁금했습니다.

## 해결한 점
- 화면에 보이는 Stock와 실제 Stock를 `fetchStock`함수로 연결해서 해결했습니다.

- 결합도가 높아지는 문제를 다음과 같이 해결했습니다.
기존의 코드에는 다음가 같은 문제가 있었다고 생각했습니다.
1. 1번 함수가 2번 함수에서 사용하고 2번 함수가  3번 함수에서 연달아 사용하게 되는 문제
2. 하나의 함수가 2가지 역할을 수행하는 문제
3. for 반복문 내부에서 checkStock 를 해주어서  `guard let _ = ...` 와 같이 불필요한 변수를 `_`로 처리하여 가독성이 떨어지는 문제
4. +/- 의 역할을 메소드 주입을 통해 구현한 점

**기존 코드**
```swift
// FruitStore.swift
func checkStock(
    of fruit: Fruit,
    in amount: Int,
    by operation: (Int, Int) -> Int
  ) -> Result<Int, MakeJuiceError> {
  guard let item = self.stock[fruit] else {
    return .failure(.notExistFruit)
  }
  guard operation(item, amount) >= 0 else {
    return .failure(.outOfStock)
  }
  return .success(item)
}
  
func changeStock(of fruit: Fruit, in amount: Int, by operation: (Int, Int) -> Int) {
  guard let item = self.stock[fruit] else {
    return
  }
  self.stock[fruit] = operation(item, amount)
}

// JuiceMaker.swift
private func canMake(_ juice: Juice) -> Bool {
  for (fruit, amount) in juice.recipe {
    guard let _ = try? self.fruitStore.checkStock(of: fruit, in: amount, by: -).get() else {
      return false
    }
  }
  return true
}
  
func make(_ juice: Juice) {
  guard self.canMake(juice) else {
    return
  }
  for (fruit, amount) in juice.recipe {
    self.fruitStore.changeStock(of: fruit, in: amount, by: -)
  }
}
```
기존 코드의 FrutiStore에서 `checkStock()`에는 stock 내부에 해당 과일이 없는 경우와 연산 결과가 재고 개수보다 작거나 같은 지를 확인하는 두가지 기능을 수행하고 있고

`changeStock()`는 주입을 통해 +/- 연산을 수행하는 문제와 stock 내부에 해당 과일이 있는 지도 확인하는 문제가 있었습니다.  

JuiceMaker에서는 FruitStore의 `checkStock()`를 가져다가 재료에 해당하는 과일들의 재고가 충분한지를 `canMake()`를 통해 해줘서 JuiceMaker는 만드는 역할을 하는 객체이지만 체크하는 기능까지 담당하는 문제가 있었습니다.

---

**개선된 코드**
```swift
// FruitStore.swift
func checkStock(of fruits: [Fruit: Int]) -> Bool {
  return fruits.keys
    .filter { self.stock.keys.contains($0) }
    .count == fruits.count
}

func canMake(_ fruits: [Fruit: Int]) -> Bool {
  return self.stock
    .merging(fruits) { $0 - $1 }
    .filter { $0.value < Int.zero }
    .count == Int.zero
}

func consumeStock(of fruits: [Fruit: Int]) {
  self.stock.merge(fruits) { $0 - $1 }
}

// JuiceMaker.swift
func make(_ juice: Juice) -> Result<Juice, MakeJuiceError> {
  let recipe = juice.recipe
  guard self.fruitStore.checkStock(of: recipe) else {
    return .failure(.notExistFruit)
  }
  guard self.fruitStore.canMake(recipe) else {
    return .failure(.outOfStock)
  }
  self.fruitStore.consumeStock(of: recipe)
  return .success(juice)
}
```
개선된 코드에서는 재고가 충분한지를 판단하는 `canMake()`역할을 FruitStore로 옮기고, 

두가지 기능을 수행하던 `checkStock()`는 해당 과일이 내부에 있는 지만 확인하도록 변경했습니다.

`changeStock()`은 `consumeStock()`로 주입을 통한 방식에서 소비한다는 의미를 더 명시할 수 있도록 네이밍을 변경하고 재고에서 개수를 줄여주는 역할만 해야한다고 생각하여 개선했습니다.

## 조언 받고 싶은 점 

- stockVC 와 같이 ViewController의 약자를 사용해도 되는지 궁금합니다

- `strawberryCountLabel.text`안에 `fruitStore.stock[strawberry]`를 String으로 형변환 하여 넣어야 하는데 만약 `fruitStore.stock[]`가 형변환 실패시 0을 넣어준다는 의미로 nil병합 연산자를 사용했는데 만약 이부분이 이상하다면 어떻게 하는것이 좋을까요 
```swift
  func fetchStock() {
    let fruitStore = FruitStore.shared
    strawberryCountLabel.text = "\(fruitStore.stock[.strawberry] ?? 0)"
    bananaCountLabel.text = "\(fruitStore.stock[.banana] ?? 0)"
    pineappleCountLabel.text = "\(fruitStore.stock[.pineapple] ?? 0)"
    kiwiCountLabel.text = "\(fruitStore.stock[.kiwi] ?? 0)"
    mangoCountLabel.text = "\(fruitStore.stock[.mango] ?? 0)"
  }
```

- `didTapOrderButton`함수에서 `success`와 `failure`에서 얼럿을 생성하는 함수 길이가 너무 길어진것 같아서 이 함수를 분리해야할지 아니면 그대로 사용할지 조언을 구하고 싶습니다.

```swift
 @IBAction func didTapOrderButton(_ sender: UIButton) {
    guard let juice = self.selectJuice(sender) else {
      return
    }
    let juiceMaker = JuiceMaker()
    let result = juiceMaker.make(juice)
    switch result {
    case .success(let juice):
      let okAction = UIAlertAction(title: AlertSetting.ok, style: .default)
      let alert = self.createAlert(
        title: AlertSetting.notice,
        message: juice.description,
        preferredStyle: .alert,
        actions: [okAction]
      )
      self.present(alert, animated: true)
    case .failure(let error):
      let okAction = UIAlertAction(title: AlertSetting.yes, style: .default) { _ in
        guard let stockVC = self.storyboard?.instantiateViewController(withIdentifier: "stockVC")
        else {
          return
        }
        self.present(stockVC, animated: true)
      }
      let noAction = UIAlertAction(title: AlertSetting.no, style: .destructive)
      let alert = self.createAlert(
        title: AlertSetting.notice,
        message: error.errorDescription,
        preferredStyle: .alert,
        actions: [noAction, okAction]
      )
      self.present(alert, animated: true)
    }
    self.fetchStock()
  }
}
```

Commit:
feat:  ViewController IBOutlet, IBAction 연결

Juice Order Button 및 Fruit Count Label IBOutlet과 IBAction 연결

Taeangel committed:
feat: FruitViewController 생성 및 IBOutlet, IBAction 연결

Fruit Count Label 및 각각 Stepper 연결, FruitViewController 생성

Commit:
refactor: FruitStore, JuiceMaker 함수 수정

JuiceMaker의 canMake를 FruitStore의 함수로 이동하였고 기능을 분리함
Dictionary의 merging, merge를 사용하여 로직을 수정

Taeangel committed:
feat: createAlertController 생성, 폴더 분리

alertcontriller 를 만들어주는 함수 생성, AppDelegate, SceneDelegate 파일을 용도에 맞게 분리

Commit:
feat: ViewController에 fetchStock 함수 생성

각 과일의 재고 개수를 Label 과 연결하는 fetchStock 함수 구현, viewDidLoad에 fetchStock 호출
FruitStore를 Singleton으로 변경

Taeangel committed:
feat: selectJuice 함수 생성, Juice타입에 rawValue 설정

sender를 받아서 쥬스 타입을 반환해 주는 selectJuice 함수 생성, Juice타입에 rawValue 설정

Commit:
feat: AlertSetting 열거 타입 생성 및 didTapOrderButton 기능 구현

Alert에서 표시될 문자를 AlertSetting 네임스페이스로 생성
재고 수정 버튼 클릭시 FristViewController 연결된 Segue를 통해 화면 전환
사용자 주문 버튼이 클릭되면 선택된 juice를 JuiceMaker의 make 함수를 통해 처리하고 결과에 따른 Alert 표시
성공 시 성공 Alert 표시, 실패 시 실패 Alert 을 표시하고 예 버튼을 클릭시 재고 수정 화면으로 Modal 방식으로 이동

Commit:
refactor: ViewController의 누락된 final 키워드 추가

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -7,9 +7,10 @@
 
 final class FruitStore {
-  private var stock = [Fruit: Int]()
+  static let shared = FruitStore()
```
이게 싱글톤인 이유가 있나요? 싱글톤을 사용하는 이유와 장점, 단점이 무엇일까요?

Taeangel replied:
FruitStore라는 하나의 객체를 공유하기 때문에 singleton을 사용했습니다. 

## 싱글톤의 장단점 
### 장점
1. 메모리 낭비를 방지할 수 있다.
2. 하나의 객체를 공유해야하는 경우 (예: 오디오 시스템 관리자, 프린트 시스템 관리자, 로그 기록, 파일 관리자)
### 단점 
1. 생성자를 private 했기 때문에 객체지향 프로그래밍의 상속을 받아 쓰는 장점을 포기하게 된다.
2. 의존성이 생기게 되어 결합도가 증가(SOLID - OCP 위반) 하게된다. 따라서, 단위 테스트하기 어렵게 만든다.
3. 전역으로 되어있어 아무 객체나 접근할 수 있기 때문에 객체의 상태를 변경해버릴 수도 있다. 따라서, 상태를 가지지 않는 클래스에 사용해야한다. (무상태 객체 혹은 설계상 유일해야하는 시스템 컴포넌트)

jungseungyeo replied:
싱글톤의 장점, 단점을 잘 알고 계신거 같아요 ㅎㅎ 여기서는 싱글톤을 사용하지 않으면 좋을거같아요. 싱글톤을 사용하면 메모리에 한번 올라가면 내려올수가 없어요. 이 모델 객체가 올리는 것은 좋아보이지 않아요.

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -21,24 +22,20 @@ final class FruitStore {
-    return .success(item)
+  func checkStock(of fruits: [Fruit: Int]) -> Bool {
+    return fruits.keys
+      .filter { self.stock.keys.contains($0) }
```
filter 사용하는거 너무 좋네요 ㅎㅎ 함수형 언어의 특징을 잘 이용하셨군요 ㅎㅎ

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -7,9 +7,18 @@
+  case strawberryBanana = "딸바"
+  case mangoKiwi = "망키"
+  
+  var description: String {
```
description을 생성하여 만드는 코드 너무 좋습니다. 다만 이것을 사용하시려면 CustomStringConvertible을 사용해보시면 좋을거같아요.

Taeangel replied:
```swift
enum Juice: String {
  case strawberry = "딸기"
  case banana = "바나나"
  case pineapple = "파인애플"
  case kiwi = "키위"
  case mango = "망고"
  case strawberryBanana = "딸바"
  case mangoKiwi = "망키"
  
  var description: String {
    return "\(self.rawValue) 쥬스 나왔습니다! 맛있게 드세요!"
  }
```
에서
```swift
extension Juice: CustomStringConvertible {
  var description: String {
    return "\(self.rawValue) 쥬스 나왔습니다! 맛있게 드세요!"
  }
}
```
으로 CustomStringConvertible 프로토콜 채택하였습니다.
표시하는 방법으로는 첫번째 방법으로는 문자열 보간법을 사용한`"\(juice)"` 방식과 두번째 방법으로는 String의 describing 생성자를 활용하는  `String(describing: juice)` 이 두가지가 있는데 저희는 `"\(juice)"`으로 수정하였습니다.

seu0313 replied:
`convertStockToString()` 함수를 통해 `fetchStock()`를 변경하다가 문자열 보간법으로 사용하는 것보다 String으로 구현하는 것이 더 좋지 않을까라고 생각했고 `String(describing: juice)` 변경했습니다.

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -7,23 +7,17 @@
+      return .failure(.outOfStock)
     }
+    self.fruitStore.consumeStock(of: recipe)
+    return .success(juice)
```
잘 리펙토링 된 거 같아요 너무 좋습니다.

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -6,13 +6,111 @@
+    mangoCountLabel.text = "\(fruitStore.stock[.mango] ?? 0)"
+  }
+  
+  func createAlert(
```
네이밍이 createAlert이라는 네이밍은 좋지 않을거 같아요. 인스던스를 생성하는 느낌이예요. 네이밍을 수정하면 좋을거 같아요. 
presentAlert, showAlert등 행위를 나타내는 네이밍이면 좋을거같아요.

seu0313 replied:
AlertSetting 으로 빼주면서 createAlert를 presentAlert로 네이밍을 변경했습니다!

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/AlertSetting.swift:
```diff
@@ -0,0 +1,15 @@
+
+import Foundation
+
+enum AlertSetting {
```
이 값이 존재한다면 따로 Alert을 컨트롤하는 객체가 존재하면 좋을거같아요.

seu0313 replied:
AlertSetting에 타입 메서드로 AlertController를 만드는 역할을 ViewController에서 분리했습니다! 😃

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -6,13 +6,111 @@
   }
   
-  
+  @IBAction func didTapOrderButton(_ sender: UIButton) {
```
함수가 너무 긴거 같아요. 역할 별로 나누어서 작성해보면 좋을거같아요. 기능 하나씩 분리하는 솔리드 원칙의 단일책임의 원칙을 적용해보면 좋을거같아요.

seu0313 replied:
성공시 생성될 Alert 코드와 실패시 생성될 코드를 `presentSuccessAlert()`, `presentFailureAlert()` 함수로 기능을 분리했습니다!

jungseungyeo reviewed (changes requested):
링고, 태엔젤 고생하셨어요. 

코드가 점점 좋아지는 것이 보이네요 ㅎㅎ 함수형 언어의 대해서 이해하고 작성한 느낌이 들어서 좋았습니다.

### 고민한 점, 해결한 점

`이 부분은 step에서 제가 수정을 요청한 부분이였네요. 매우 잘 해결하셨어요. 특히 중복코드도 제거가 되었고 함수 하나하나가 다 존재 이유를 알 수 있어서 좋았습니다. 또 filter, merge 와 같이 함수형 언어의 특징을 잘 사용하여 좋았습니다.`

### 조언 받고 싶은 점

### stockVC 와 같이 ViewController의 약자를 사용해도 되는지 궁금합니다

> 개인적으로는 VC와 ViewController의 둘다 사용해도 괜찮다고 생각을 하고 있으나 통상적으로 ViewController를 다 붙여주는 편입니다. 약식 보다는 풀 네임을 적어 명확하게 네이밍을 표시하는 것이 대부분의 개발자들이 선호하는 방식으로 이해하고 있습니다.

### strawberryCountLabel.text안에 fruitStore.stock[strawberry]를 String으로 형변환 하여 넣어야 하는데 만약 fruitStore.stock[]가 형변환 실패시 0을 넣어준다는 의미로 nil병합 연산자를 사용했는데 만약 이부분이 이상하다면 어떻게 하는것이 좋을까요

> 0을 넣어주어도 상관을 없을거 같지만 더 좋은 방법이 있을거같아요. 값을 가져오지 못하는 경우 throw을 던져 주는 함수가 있다면 바로 해결이 될 거같네요 바로 가져와서 넣어주기 보다는 값을 배출하는 함수를 이용하는 방식이 있을거같아요.

``` swift
  func fetchStock() {
    let fruitStore = FruitStore.shared
    strawberryCountLabel.text = "\(fruitStore.stock[.strawberry] ?? 0)"
    bananaCountLabel.text = "\(fruitStore.stock[.banana] ?? 0)"
    pineappleCountLabel.text = "\(fruitStore.stock[.pineapple] ?? 0)"
    kiwiCountLabel.text = "\(fruitStore.stock[.kiwi] ?? 0)"
    mangoCountLabel.text = "\(fruitStore.stock[.mango] ?? 0)"
  }
```

### didTapOrderButton함수에서 success와 failure에서 얼럿을 생성하는 함수 길이가 너무 길어진것 같아서 이 함수를 분리해야할지 아니면 그대로 사용할지 조언을 구하고 싶습니다.

> 제 의견도 함수가 너무 긴거 같아요. 함수가의 역할을 나눌 수 있을거 같아요. 코멘트를 보시고 적용하시면 좋을거같아요.

Lingo, Taeangle 앱을 완성가 좋습니다. 다만 역할의 분리가 아직 미흡한 부분이 조금 남아있는거 같아요. 이 부분 수정해주시면 좋을거같아요 :) 하시다가 궁금한 점이 있다면 언제든지 물어봐주세요 :)

Taeangel committed:
refactor: CustomStringConvertible를 채택

Juice타입에서 CustomStringConvertible를 채택하여 수정

Commit:
refactor: ViewController의 createAlert 이동, presentAlert 네이밍 변경

createAlert를 presentAlert로 네이밍 변경
ViewController에 있던 presentAlert을 AlertSetting에서 타입 메서드로 이동 및 변경

Taeangel committed:
refactor: presentSuccessAlert, presentFailureAlert, handleOkAction 함수를 생성해서 분리

didTapOrderButton에서 성공시 보여줄 Alert과 실패시 보여줄 Alert를 함수로 따로 분리
실패시 보여줄 alert에서 okAction의 handler를 따로 함수로 분리

seu0313 commented:
선택된 쥬스 버튼에 따라 쥬스를 반환하는 `selectJuice()` 함수에서
**기존** 
```swift
func selectJuice(_ sender: UIButton) -> Juice? {
  switch sender {
  case strawberryBananaButton:
    return .strawberryBanana
  case mangoKiwiButton:
    return .mangoKiwi
  case strawberryButton:
    return .strawberry
  case bananaButton:
    return .banana
  case pineappleButton:
    return .pineapple
  case kiwiButton:
    return .kiwi
  case mangoButton:
    return .mango
  default:
    return nil
  }
}

```
switch문을 사용하여 처리하던 방식에서

**게선**
```swift
func selectJuice(_ sender: UIButton) -> Juice? {
  guard let buttonTitle = sender.titleLabel?.text,
        let juiceName = buttonTitle.components(separatedBy: " ").first
  else {
    return nil
  }
  return Juice(rawValue: juiceName)
}
```
해당 버튼의 타이틀에 있는 값을 가져와서 Juice(rawValue:)로 개선 해봤는데 사용해도 되는지 궁금합니다.

Commit:
refactor: ViewController의 selectJuice 함수 수정

selectJuice 함수에서 switch문을 사용해서 클릭된 버튼에 따라 Juice를 반환하던 방식에서
해당 버튼의 Title을 사용하여 Juice타입을 생성하여 리턴하도록 수정

Taeangel committed:
refactor: convertStockToString 함수생성

Int?값을 바인딩 해서 성공하면 String으로 반환하고 값을 가져오지 못하는 경우 0을 반환하는 convertStockToString 함수를 생성

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -96,4 +75,35 @@ private extension ViewController {
     mangoCountLabel.text = "\(fruitStore.stock[.mango] ?? 0)"
   }
+  
+  private func presentSuccessAlert(_ juice: Juice) {
```
이거는 alert을 두 함수로 나눌 필요는 없는거 같아요. alert을 노출하는 인터페이스는 하나로 통일하는 방식이 좋을거같아요.

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -6,13 +6,98 @@
+    return String(fruitAmount)
+  }
+  
+  func presentSuccessAlert(_ juice: Juice) {
```
alert 인터페이스는 하나로 만들면 좋을거같아요. successalert, failurealert으로 나눌 필요는 없을거같아요.

seu0313 replied:
두 개로 나눴던 함수를 `presentAlert()`로 파라미터로 타입이 `String?`인 message를 받도록 인터페이스를 통합했습니다.
```swift
func presentAlert(message: String?, actions: [UIAlertAction]) {
    let alert = AlertSetting.createAlertController(
      title: AlertSetting.notice,
      message: message,
      preferredStyle: .alert,
      actions: actions
    )
    self.present(alert, animated: true)
  }
``` 
프로젝트 요구조건에서 실패시에는 `예/아니오`로 구현하라고 되어있고 성공시에는 `예/아니오`로 할 수 없기 때문에 actions 파라미터에`UIAlertAction`을 선언하여 배열로 넣을 수 있도록 했습니다.

jungseungyeo reviewed (changes requested):
링고, 태엔젤 리뷰 답변 잘 하셨어요. 몇가지 수정을 해주셨으면 좋겠습니다. 모델의 싱글톤을 없애고 alert의 인터페이스를 하나로 변경하면 좋을거같아요. 

모델을 싱글톤의 장 단점을 잘 알고 있지만, 싱글톤을 쓰는것을 추천하지 않는 이유는 해당 객체를 싱글톤으로 사용 할 경우 메모리를 올리면 해제가 되지 않아요. 그 말은 deinit이 호출을 할 수 없다는 거라서 싱글톤을 제거 하면 좋을거같아요.

alert 인터페이스는 성공, 실패를 나눠서 하기 보다는 하나의 인터페이스를 만들고 성공 실패 시점에서 인터페이스에 맞게 주면 좋을거같아요. :) 

이 부분만 수정하면 될거같아요.

Commit:
refactor: FruitStore 싱글톤 수정, ViewController 네이밍 변경

FruitStore, JuiceMaker의 싱글톤을 제거
기존 Main StockViewController의 storyboard id를 StockVC에서 StockViewController로 변경
ViewController와 FruitViewController의 이름을 JuiceViewController, StockViewController로 변경

Taeangel committed:
refactor: alert interface통합

성공시 보여줄 alert과 실패시 보여줄 alert의 인터페이스 통합
alert과 관련된 함수이름을 상황에 맞게 변경

Taeangel commented:
싱글톤을 제거하고 JuiceMaker를 JuiceViewController에서 전역변수로 선언해서 수정하였습니다.

Taeangel committed:
style: 누락된 self 추가, 하드코딩 수정

Commit:
refactor: IBOutlet, IBAction 접근 제어 수정

JuiceViewController의 juiceMaker와 IBOutlet, IBAction가 외부에서 접근할 수 없도록 internal에서 private로 변경

Commit:
style: JuiceViewController의 누락된 줄바꿈 추가

jungseungyeo commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -7,10 +7,9 @@
   private(set) var stock = [Fruit: Int]()
   
-  private init() {
+  init() {
```
좋습니다 :)

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/JuiceViewController.swift:
```diff
@@ -8,28 +8,28 @@ import UIKit
-  @IBOutlet weak var pineappleCountLabel: UILabel!
-  @IBOutlet weak var kiwiCountLabel: UILabel!
-  @IBOutlet weak var mangoCountLabel: UILabel!
+  @IBOutlet private weak var strawberryCountLabel: UILabel!
```
접근제어자 붙여주는거 너무 좋아요 :)

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/JuiceViewController.swift:
```diff
@@ -71,25 +74,12 @@ private extension JuiceViewController {
-  func presentSuccessAlert(_ juice: Juice) {
-    let okAction = UIAlertAction(title: AlertSetting.ok, style: .default)
-    let alert = AlertSetting.presentAlert(
+  func presentAlert(message: String?, actions: [UIAlertAction]) {
```
너무 좋아요 ㅎㅎ 완벽합니다:)

jungseungyeo reviewed (approved):
링고, 태엔젤 고생하셨어요 다음 step 진행하면 좋을거같아요 :)
